include/pybind11/cast.h

```diff
@@ -858,7 +882,8 @@ struct polymorphic_type_hook : public polymorphic_type_hook_base<itype> {};
 PYBIND11_NAMESPACE_BEGIN(detail)
 
 /// Generic type caster for objects stored on the heap
-template <typename type> class type_caster_base : public type_caster_generic {
+template <typename type> class type_caster_base : public type_caster_generic,
+                                                  protected make_constructor {
```

Comment:
Uuuu, multiple inheritance. I guess it should be fine (there's nothing virtual about these casters, right?), but there's really no other way to get this done? :-/

Nitpick: Let's name it `make_constructor_base` or `make_constructor_mixin`, maybe, to make clear it's a class, rather than a function. (I guess that's obvious when used as base type, but still, `make_constructor` sounds like a function to me.)

Reply:
> Uuuu, multiple inheritance. I guess it should be fine (there's nothing virtual about these casters, right?), but there's really no other way to get this done? :-/

Within pybind11, i see no reason to actually have this be a static member function? So just adding these as inline functions to `detail` might work as well and avoid MI.

But we might be breaking the semi-public interface of `type_caster_base` (since it's only protected).

Reply:
Yes, non-virtual, so multiple inheritance is straightforward/safe.
But let me try something else, a little less conservative ... hang on.

Reply:
Yes, maybe this is the most interface-preserving change you can make, if we assume users subclassing `type_caster_base` would depend on this protected member.

And duplicating
```c++
template <typename T>
static auto make_copy_constructor(const T *x) -> decltype( detail::make_copy_constructor(x)) {
    return detail::make_copy_constructor(x);
}
```
just to avoid a mostly-safe type of MI is also quite silly.

Reply:
Curious to see what you still have in mind, though!
Otherwise, I'm fine with this, except for maybe the suggested name change, if you agree on that.